keras_nlp/models/bloom/bloom_attention.py

```diff
@@ -0,0 +1,193 @@
+# Copyright 2023 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import math
+
+from keras_nlp.backend import keras
+from keras_nlp.backend import ops
+
+
+class BloomAttention(keras.layers.Layer):
+    def __init__(
+        self,
+        num_heads,
+        dropout=0.0,
+        kernel_initializer="glorot_uniform",
+        bias_initializer="zeros",
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.num_heads = num_heads
+        self.dropout = dropout
+        self.kernel_initializer = kernel_initializer
+        self.bias_initializer = bias_initializer
+
+    def build(self, inputs_shape):
+        _, seq_length, hidden_dim = inputs_shape
+
+        self.head_dim = hidden_dim // self.num_heads
+
+        # Layer-wise attention scaling
+        self.inv_norm_factor = 1.0 / math.sqrt(self.head_dim)
+
+        self._query_key_value_dense = keras.layers.Dense(
+            3 * hidden_dim,
+            kernel_initializer=self.kernel_initializer,
+            bias_initializer=self.bias_initializer,
+            dtype=self.dtype_policy,
+            name="query_key_value_dense",
+        )
+        self._query_key_value_dense.build(inputs_shape)
+
+        self._out_dense = keras.layers.Dense(
+            hidden_dim,
+            kernel_initializer=self.kernel_initializer,
+            bias_initializer=self.bias_initializer,
+            dtype=self.dtype_policy,
+            name="out_dense",
+        )
+        self._out_dense.build(inputs_shape)
+
+        self._dropout = keras.layers.Dropout(
+            rate=self.dropout, dtype=self.dtype_policy, name="dropout"
+        )
+        self._softmax = keras.layers.Softmax(
+            dtype=self.dtype_policy, name="softmax"
+        )
+
+        self.built = True
+
+    def _split_heads(self, fused_qkv):
+        batch_size, seq_length, _ = ops.shape(fused_qkv)
+        fused_qkv = ops.reshape(
+            fused_qkv,
+            (batch_size, seq_length, self.num_heads, 3, self.head_dim),
+        )
+        return fused_qkv[..., 0, :], fused_qkv[..., 1, :], fused_qkv[..., 2, :]
+
+    @staticmethod
+    def _build_alibi_tensor(seq_length, num_heads):
```

Comment:
Probably best as a follow up PR, but this implantation of alibi is the same as other models, we might consider splitting this into a standalone layer (can use `keras_nlp.layers.RotaryEmbedding` as a template).

Do you know if there is any variation in how bloom is applying alibi compared to other models?

Reply:
there is no difference between bloom alibi tensor and alibi tensor in other models. the only difference is that bloom uses `alibi_bias_max` that equals to 8 always which is the same in the original paper. I added `alibi_bias_max` to the method so any one who wants to use alibi tensor can change it (when we split it as a standalone layer).

I got the naming `alibi_bias_max` from this huggingface implementation https://huggingface.co/mosaicml/mpt-7b/blob/ada218f9a93b5f1c6dce48a4cc9ff01fcba431e7/attention.py#L330

regarding creating a standalone layer, what I understood is that we need to implement a layer that only generates the alibi tensor not a whole attention layer that uses alibi, is that right?

Reply:
Yeah, not sure what the best design is TBH, but that seems the most modular so that is what we are going with right now.

Make a later that generates the alibi tensors that can be used in any attention layer built on top.